drivers/serial/uart_s32_linflexd.c

```diff
@@ -0,0 +1,531 @@
+ *
+ * @return 0
+ */
+static int uart_s32_init(const struct device *dev)
+{
+	const struct uart_s32_config *config = dev->config;
+	uint8_t key;
+	int err;
+
+	err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (err < 0) {
+		return err;
+	}
+
+	/* disable interrupts */
+	key = irq_lock();
+
+	/* Setup UART init structure */
+	Linflexd_Uart_Ip_Init(config->instance, config->uart_hw_cfg);
+
+	/* restore interrupt state */
+	irq_unlock(key);
+
+	return 0;
+}
+
+static const struct uart_driver_api uart_s32_driver_api = {
+	.poll_in	  = uart_s32_poll_in,
+	.poll_out	  = uart_s32_poll_out,
+	.err_check	  = uart_s32_err_check,
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill	  = uart_s32_fifo_fill,
+	.fifo_read	  = uart_s32_fifo_read,
+	.irq_tx_enable	  = uart_s32_irq_tx_enable,
+	.irq_tx_disable   = uart_s32_irq_tx_disable,
+	.irq_tx_ready	  = uart_s32_irq_tx_ready,
+	.irq_rx_enable	  = uart_s32_irq_rx_enable,
+	.irq_rx_disable   = uart_s32_irq_rx_disable,
+	.irq_rx_ready	  = uart_s32_irq_rx_ready,
+	.irq_err_enable   = uart_s32_irq_err_enable,
+	.irq_err_disable  = uart_s32_irq_err_disable,
+	.irq_is_pending   = uart_s32_irq_is_pending,
+	.irq_update	  = uart_s32_irq_update,
+	.irq_callback_set = uart_s32_irq_callback_set,
+#endif	/* CONFIG_UART_INTERRUPT_DRIVEN */
+
+};
+
+#define UART_S32_NODE(n)	DT_NODELABEL(uart##n)
+
+#define _GET_HW_CFG(n, func_name)	Linflexd_Uart_Ip_xHwConfigPB_##n##_##func_name
+#define GET_HW_CFG(n, func_name)	_GET_HW_CFG(n, func_name)
+#define LINFLEXD_GET_HW_CFG(n)		GET_HW_CFG(n, S32CT_FUNC_GROUP_UPPER)
+
+#if defined(CONFIG_UART_INTERRUPT_DRIVEN)
+#define UART_S32_INTERRUPT_DEFINE(n, isr_handler, parameter)	\
+	do {							\
+		IRQ_CONNECT(DT_IRQN(UART_S32_NODE(n)),		\
+			DT_IRQ(UART_S32_NODE(n), priority),	\
+			isr_handler,				\
+			parameter,				\
+			DT_IRQ(UART_S32_NODE(n), flags));	\
+								\
+		irq_enable(DT_IRQN(UART_S32_NODE(n)));		\
+	} while (0)
+
+#else
+#define UART_S32_INTERRUPT_DEFINE(n, isr_handler, parameter)
+#endif
+
+#define UART_S32_INIT_DEVICE(n)							\
+	PINCTRL_DT_DEFINE(UART_S32_NODE(n));					\
+	static struct uart_s32_data uart_s32_data_##n;				\
+	static const struct uart_s32_config uart_s32_config_##n = {		\
+		.instance = n,							\
+		.base = (LINFLEXD_Type *)DT_REG_ADDR(UART_S32_NODE(n)),		\
+		.uart_hw_cfg = (Linflexd_Uart_Ip_UserConfigType *)&LINFLEXD_GET_HW_CFG(n),\
+		.pincfg = PINCTRL_DT_DEV_CONFIG_GET(UART_S32_NODE(n)),		\
+	};									\
+	static int uart_s32_init_##n(const struct device *dev)			\
+	{									\
+		UART_S32_INTERRUPT_DEFINE(n, Linflexd_Uart_Ip_IRQHandler, n);	\
+										\
+		return uart_s32_init(dev);					\
+	}									\
+	DEVICE_DT_DEFINE(UART_S32_NODE(n),					\
+			&uart_s32_init_##n,					\
+			NULL,							\
+			&uart_s32_data_##n,					\
+			&uart_s32_config_##n,					\
+			PRE_KERNEL_1,						\
+			CONFIG_KERNEL_INIT_PRIORITY_DEVICE,			\
+			&uart_s32_driver_api);
+
+#if DT_NODE_HAS_STATUS(UART_S32_NODE(0), okay)
```

Comment:
The device driver needs to keep track of the hw peripheral instance number as required by the HAL. Using the `DT_INST` approach is indeed less verbose and more compact, but it would require to store the instance number in DT. I'm ok with either way, as long as we stick to it with the rest of the drivers to come.

Reply:
What you did is the correct way to handle it.  Its unfortunate that the HAL is based on a index.

Reply:
Thanks galak for clarification